app/donu/Main.hs

```diff
@@ -1,72 +1,108 @@
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TemplateHaskell #-}
 module Main ( main ) where
 
-import Control.Monad.IO.Class ( liftIO )
-import Control.Exception      ( try, SomeException )
+import Control.Monad.State
+
+import           Control.Applicative ((<|>))
+import           Control.Lens.TH
+import qualified Web.ClientSession as ClientSession
 
 import qualified Data.Text as Text
 import qualified Data.Aeson as JS
+import           Data.Maybe (fromMaybe)
 
-import Language.ASKEE
+import qualified Snap
 
-import Schema
+import           Language.ASKEE
+import qualified Language.ASKEE.Exposure.Interpreter as Exposure
+import           Language.ASKEE.Exposure.GenLexer (lexExposure)
+import           Language.ASKEE.Exposure.GenParser (parseExposureStmts)
 
-import qualified Snap.Core as Snap
-import           Snap.Http.Server ( quickHttpServe )
+import           Schema
+import           ExposureSession
+import qualified Language.ASKEE.Exposure.Syntax as Exposure
 
-main :: IO ()
-main = 
-  do  initStorage
-      quickHttpServe $
-        do  Snap.route [ ("/help", showHelp)
-                       , ("/", endpoint)
-                       ]
+-- Snaplet Definition ---------------------------------------------------------
+
+newtype Donu sim = Donu
+ { _exposureSessions :: Snap.Snaplet (ExposureSessionManager sim)
+   -- ^ Snaplet implementing Exposure REPL sessions
+ }
+
+makeLenses ''Donu
+
+-------------------------------------------------------------------------------
+
+type ExposureState = Exposure.Env
+type DonuApp = Donu ExposureState
+
+initDonu :: Snap.SnapletInit DonuApp DonuApp
+initDonu =
+  Snap.makeSnaplet "donu" "donu" Nothing $
+    do sess <- Snap.nestSnaplet "sessions" exposureSessions $
+                 initExposureSessionManager
+                   ClientSession.defaultKeyFile -- TODO: Do we care about this?
+                   "exposure-env-id"
+                   1800
+       Snap.addRoutes [ ("/help", showHelp)
+                      , ("/", endpoint)
+                      ]
+       return $ Donu sess
   where
     endpoint =
      do let limit = 8 * 1024 * 1024    -- 8 megs
         Snap.modifyResponse (Snap.setHeader "Access-Control-Allow-Origin" "*")
         body <- Snap.readRequestBody limit
         case JS.eitherDecode body of
           Right a ->
-            do  r <- liftIO $ try $ handleRequest a
-                case r of
-                  Right out ->
-                    do  let (code, msg) =
-                              case out of
-                                (FailureResult _) -> (400, "Error")
-                                _ -> (200, "OK")
-                        Snap.modifyResponse (Snap.setResponseStatus code msg)
-                        Snap.writeLBS (JS.encode out)
-                  Left err -> errorWith 500 (show (err :: SomeException))
+               runHandler a
+           <|> errorWith 400 "Bad request"
```

Comment:
@abakst: I'm not quite sure I preserved the intent of this error-generating code when I migrated it over to use `errorWith`. It may be worth double-checking that I did this the way you'd expect.